hamilton/function_modifiers/recursive.py

```diff
@@ -11,17 +14,21 @@
 else:
     from typing import NotRequired
 
+from pandas import DataFrame as PandasDataFrame
+from polars import DataFrame as PolarsDataFrame
+from polars import LazyFrame as PolarsLazyFrame
```

Comment:
Thanks for all the feedback @zilto!

Happy to go ahead with `registry`-approach since it seems a straightforward change / keeps it consistent with `extract_columns` that is used under the hood.

Could you confirm the following just to double-check that I understand correctly:
The change necessary is that we move away from a factory pattern with subclassing the different implementations from and changing the now abstract method `create_merge_node` into a `@singledispatch` method with the plugin extensions `h_pandas` and `h_polars` implementing it, the same as in `extract_columns`.

Reply:
`create_merge_node` I don't think is the right way to do this, but close. 

The problem here is that we have not really defined an algebra for doing the atomic operations for `with_columns` cross-dataframe and it's not actually clear it's the same:

1. Spark does it differently
2. Pandas + polars are similar actually, but lazy stuff gets in the way

So, while long-term it's a bit cleaner to come up with a set of atomic operations and delegate to each one, I think that it's a little more than we need right now.

We should be able to register `with_columns` -- we can merge stuff later, but it'll be unecessarily restrictive for now.

Narwals can help, but really this is fairly simple -- we have three implementations and they don't use the same code (yet), so unless we confidently believe that they'll all look the same, I'd prefer to duplicate code and reimplement the entirity of `with_columns`. Perhaps we can group this to use the same interface, but again, it's not clear that all dataframe implementations *will* use the same interface...